models/service.rb

```diff
@@ -53,5 +54,14 @@ class Service < Base
     has_many :activities, class_name: 'Models::Activity'
     has_many :sticky_vehicles, class_name: 'Models::Vehicle'
     has_many :quantities, class_name: 'Models::Quantity'
+
+    def duplicate_safe(options = {})
```

Comment:
I thought we decided to keep all this duplicate logic inside periodic_visits until we have the generic clone function ready so that it wouldn't be used somewhere else, do I remember wrong?

I find this function dangerous because it has periodic_visits logic in it (for example a generic duplicate function wouldn't skip `:sticky_vehicle_ids, :quantity_ids`) but it is inside service.rb (the same for vehicles, shipments)

Reply:
You are right, if I move it in periodic_visits.rb we make it harder to use this function and we will be forced to implement a better way. I will move it.

Still, I do think this was a generic function : you can have sticky_vehicle_ids and quantity_ids even if you do not have a schedule. But you are right, this would probably come from a scheduling optimization since it implies using scheduling front (which works well even if you do not provide any schedule in configuration).

Reply:
I wasn't clear enough about the thing that is not generic.

The function is not generic, not because of the existence of `sticky_vehicle_ids` or `quantity_ids` (of course we can have them or not) but it is not generic because of the fact that we are skipping these in this duplicate function :
```ruby
next if [:sticky_vehicle_ids, :quantity_ids].include?(k)
```

This "skipping" logic makes it specific in my opinion and it wouldn't be right to make it a generic `Service` class function. A generic duplicate function would duplicate the objects faithfully.

Reply:
I see. Then I might need @braktar 's help and yours : 

self.attributes iterates over every field defined within grape model. How can we rather iterate over active_hash model attributes ? 

Or maybe we can remove attributes with key **k** such that :
`key.include?("_id") || key.include?("_ids") || key.include?("y_ids")`
after model is created.

What would be the best practice here ?

Reply:
I don't know immediately but in any case, this PR is for the non-generic version, so lets move this function inside periodic_visits and we can finalize this PR. 

We shall discuss how to implement the generic version on the other ticket when we are  working on it. I think it'll be clearer how to handle all of this, one we are inside the base.rb.

I just wanted to make myself clear about why I didn't see this function a generic function that's it.

Reply:
fields with '_id' or '_ids' are sometimes part of the model.

I think that iterate over the fields define within the model would be more generic.

https://github.com/zilkey/active_hash/blob/master/lib/active_hash/base.rb
We could probably explore the possibility to use `field_names` methods

Reply:
field_names seems undefined so I will just move functions in periodic.rb and we will discuss later as @senhalil suggested

Thank you for your help :)

Reply:
@senhalil we still need one function per model (vehicle/service/shipment) because we call Model::new. Do we still want to put it in periodic_visits.rb ?

Reply:
It's normal that we need different functions for each of these, even if we move it to `periodic_visits.rb`. The reason why we want to move it to there is that they shouldn't be used. But basically what I had in mind was a function (lets call it `periodic_visits::duplicate_model`) which duplicates the necessary parts the way periodic visits needs. Lets discuss this on the meeting via screen sharing on Monday.